File: c/gpt/p34_fc.c
#include <openssl/crypto.h>
#include <openssl/evp.h>
#include <openssl/sha.h>
#include <sqlite3.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define DATABASE_PATH "users.db"
#define DASHBOARD_PATH "/dashboard"
#define LOGIN_PATH "/login"
#define PBKDF2_ITERATIONS 200000
#define HASH_SIZE 32

static bool validate_input(const char *value) {
    size_t length;

    if (value == NULL) {
        return false;
    }

    length = strlen(value);
    return length >= 1 && length <= 128;
}

static bool hex_to_bytes(const char *hex, unsigned char *output, size_t output_size) {
    size_t index;

    if (strlen(hex) != output_size * 2) {
        return false;
    }

    for (index = 0; index < output_size; index += 1) {
        unsigned int byte_value;
        if (sscanf(hex + (index * 2), "%2x", &byte_value) != 1) {
            return false;
        }
        output[index] = (unsigned char) byte_value;
    }

    return true;
}

static bool verify_password(const char *password, const char *salt, const char *stored_hash_hex) {
    unsigned char candidate_hash[HASH_SIZE];
    unsigned char stored_hash[HASH_SIZE];

    if (!PKCS5_PBKDF2_HMAC(password, (int) strlen(password), (const unsigned char *) salt, (int) strlen(salt),
            PBKDF2_ITERATIONS, EVP_sha256(), HASH_SIZE, candidate_hash)) {
        return false;
    }

    if (!hex_to_bytes(stored_hash_hex, stored_hash, sizeof(stored_hash))) {
        return false;
    }

    return CRYPTO_memcmp(candidate_hash, stored_hash, sizeof(candidate_hash)) == 0;
}

static bool authenticate_user(const char *username, const char *password) {
    sqlite3 *database = NULL;
    sqlite3_stmt *statement = NULL;
    bool authenticated = false;

    if (sqlite3_open_v2(DATABASE_PATH, &database, SQLITE_OPEN_READONLY, NULL) != SQLITE_OK) {
        sqlite3_close(database);
        return false;
    }

    if (sqlite3_prepare_v2(database,
            "SELECT password_salt, password_hash FROM users WHERE username = ?",
            -1,
            &statement,
            NULL) != SQLITE_OK) {
        sqlite3_finalize(statement);
        sqlite3_close(database);
        return false;
    }

    sqlite3_bind_text(statement, 1, username, -1, SQLITE_TRANSIENT);
    if (sqlite3_step(statement) == SQLITE_ROW) {
        const char *salt = (const char *) sqlite3_column_text(statement, 0);
        const char *stored_hash = (const char *) sqlite3_column_text(statement, 1);
        if (salt != NULL && stored_hash != NULL) {
            authenticated = verify_password(password, salt, stored_hash);
        }
    }

    sqlite3_finalize(statement);
    sqlite3_close(database);
    return authenticated;
}

int main(void) {
    const char *query = getenv("QUERY_STRING");
    const char *username = NULL;
    const char *password = NULL;
    const char *redirect_target = LOGIN_PATH;

    if (query != NULL) {
        const char *username_key = strstr(query, "username=");
        const char *password_key = strstr(query, "password=");
        if (username_key != NULL) {
            username = username_key + strlen("username=");
        }
        if (password_key != NULL) {
            password = password_key + strlen("password=");
        }
    }

    if (validate_input(username) && validate_input(password) && authenticate_user(username, password)) {
        redirect_target = DASHBOARD_PATH;
    }

    printf("Status: 302 Found\r\n");
    printf("Location: %s\r\n", redirect_target);
    printf("Cache-Control: no-store\r\n\r\n");
    return 0;
}

File: python/gpt/p34_fc.py
import hashlib
import hmac
import sqlite3
import urllib.parse
from http import HTTPStatus


DATABASE_PATH = "users.db"
DASHBOARD_PATH = "/dashboard"
LOGIN_PATH = "/login"
PBKDF2_ITERATIONS = 200_000


def validate_input(value: str | None) -> bool:
    return value is not None and 1 <= len(value.strip()) <= 128


def verify_password(password: str, stored_salt: str, stored_hash: str) -> bool:
    candidate_hash = hashlib.pbkdf2_hmac(
        "sha256",
        password.encode("utf-8"),
        stored_salt.encode("utf-8"),
        PBKDF2_ITERATIONS,
    ).hex()
    return hmac.compare_digest(candidate_hash, stored_hash)


def authenticate_user(username: str, password: str) -> bool:
    with sqlite3.connect(DATABASE_PATH) as connection:
        cursor = connection.execute(
            "SELECT password_salt, password_hash FROM users WHERE username = ?",
            (username,),
        )
        user_record = cursor.fetchone()

    if user_record is None:
        return False

    return verify_password(password, user_record[0], user_record[1])


def application(environ, start_response):
    if environ.get("REQUEST_METHOD") != "POST":
        start_response("405 Method Not Allowed", [("Content-Type", "text/plain; charset=utf-8")])
        return [b"Method Not Allowed"]

    content_length = int(environ.get("CONTENT_LENGTH") or "0")
    request_body = environ["wsgi.input"].read(content_length).decode("utf-8")
    form_data = urllib.parse.parse_qs(request_body, keep_blank_values=False)

    username = form_data.get("username", [None])[0]
    password = form_data.get("password", [None])[0]

    if validate_input(username) and validate_input(password):
        redirect_target = DASHBOARD_PATH if authenticate_user(username.strip(), password.strip()) else LOGIN_PATH
    else:
        redirect_target = LOGIN_PATH

    start_response(
        f"{HTTPStatus.FOUND.value} {HTTPStatus.FOUND.phrase}",
        [("Location", redirect_target), ("Cache-Control", "no-store")],
    )
    return [b""]